internal/verification/database/health_authority_db.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/exposure-notifications-server/internal/database"
+	"github.com/google/exposure-notifications-server/internal/verification/model"
+
+	pgx "github.com/jackc/pgx/v4"
+)
+
+// HealthAuthorityDB allows for opreations against authorized health authorities
+// for diagnosis signature verification.
+type HealthAuthorityDB struct {
+	db *database.DB
+}
+
+// New creates a HealthAuthorityDB attached to a specific database driver.
+func New(db *database.DB) *HealthAuthorityDB {
+	return &HealthAuthorityDB{db}
+}
+
+// AddHealthAuthority inserts a new HealthAuthority record into the database.
+func (db *HealthAuthorityDB) AddHealthAuthority(ctx context.Context, ha *model.HealthAuthority) error {
+	if len(ha.Keys) != 0 {
+		return fmt.Errorf("unable to insert health authority with keys, attach keys later")
+	}
+	if err := ha.Validate(); err != nil {
+		return err
+	}
+
+	return db.db.InTx(ctx, pgx.Serializable, func(tx pgx.Tx) error {
+		row := tx.QueryRow(ctx, `
+			INSERT INTO
+				HealthAuthority
+				(iss, aud, name)
+			VALUES
+				($1, $2, $3)
+			RETURNING id
+			`, ha.Issuer, ha.Audience, ha.Name)
+		if err := row.Scan(&ha.ID); err != nil {
+			return fmt.Errorf("inserting healthauthority: %w", err)
+		}
+		return nil
+	})
+}
+
+// GetHealthAuthority retrieves a HealthAuthority record by the issuer name.
+func (db *HealthAuthorityDB) GetHealthAuthority(ctx context.Context, issuer string) (*model.HealthAuthority, error) {
+	conn, err := db.db.Pool.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("acquiring connection: %w", err)
+	}
+	defer conn.Release()
+
+	row := conn.QueryRow(ctx, `
+		SELECT
+			id, iss, aud, name
+		FROM
+			HealthAuthority
+		WHERE
+			iss = $1`, issuer)
+
+	var ha model.HealthAuthority
+	if err := row.Scan(&ha.ID, &ha.Issuer, &ha.Audience, &ha.Name); err != nil {
+		return nil, err
+	}
+
+	haks, err := db.GetHealthAuthorityKeys(ctx, &ha)
+	if err != nil {
+		return nil, err
+	}
+	ha.Keys = haks
+
+	return &ha, nil
+}
+
+func (db *HealthAuthorityDB) AddHealthAuthorityKey(ctx context.Context, ha *model.HealthAuthority, hak *model.HealthAuthorityKey) error {
+	if err := hak.Validate(); err != nil {
```

Comment:
It's probably fine, but both ha and hak could be nil, in which case this would panic